### ISSUE TITLE
VisibilityTracker & banner ads

### ISSUE THREAD

**Issue by stevenhaggerty** [AppLovin/AppLovin-MAX-SDK-Android#226]
Hi,

I'm testing mediation providers before I contemplate shifting all our apps ad provision.

We're getting some very low eCPM numbers ($0.18 for tier 1 countries) for banner ads.  We used to sing around $1.
Our rewarded/inters eCPM's are very good.

I order to see if our implementation is correct I've enabled verbose logging and I'm getting some suspicious messages from the VisibililityTracker.

Such as:

2022-03-07 17:43:43.097 3626-3626/? D/AppLovinSdk: [ViewabilityTracker] Checking visibility...
2022-03-07 17:43:43.097 3626-3626/? E/AppLovinSdk: [ViewabilityTracker] View is hidden
2022-03-07 17:43:43.103 3626-3626/? E/AppLovinSdk: [ViewabilityTracker] View is not in top activity's view hierarchy
2022-03-07 17:43:43.103 3626-3626/? D/AppLovinSdk: [ViewabilityTracker] Returning flags: 100000010

I get messages like this in my app and even when testing AppLovin banner ads by using the MediationDebugger.

Is this correct? The view is definitely not hidden and it definitely in the hierarchy. Is this affecting our eCPM?

Also:

Chartboost mediation only shows 1 banner ever - after the first show all other banner shows fail with:
AdView failed "Default" to show with error: BANNER_VIEW_IS_DETACHED

These are apps that have been on GooglePlay for years, these banner views are never detached.

I've raised this with support, but I get nothing but a ticket number, never any follow up. Ticket #12367.
I can break this Chartboost issue off into a separate post if you like. 

Thanks,

Steve.

**Comment by chauduyphanvu** [comment]
@stevenhaggerty

```
2022-03-07 17:43:43.097 3626-3626/? D/AppLovinSdk: [ViewabilityTracker] Checking visibility...
2022-03-07 17:43:43.097 3626-3626/? E/AppLovinSdk: [ViewabilityTracker] View is hidden
2022-03-07 17:43:43.103 3626-3626/? E/AppLovinSdk: [ViewabilityTracker] View is not in top activity's view hierarchy
2022-03-07 17:43:43.103 3626-3626/? D/AppLovinSdk: [ViewabilityTracker] Returning flags: 100000010
```

For the above log, that means either your MaxAdView or one of its ancestor views is not visible (see https://developer.android.com/reference/android/view/View#isShown() - `View.VISIBLE`). Also make sure the view is in the topmost Activity's view hierarchy.

For the Chartboost issue, I was informed it's a bug in their SDK and should be fixed in the upcoming release (8.4.0). Feel free to give that a try when it's out. Thanks.

Vu Chau
AppLovin Team

**Comment by tzachyrm** [comment]
@chauduyphanvu we have a similar issue with `ViewabilityTracker`, probably due to the way we render the banners:

As an optimization, we create a detached view and add it to the hierarchy only after the first banner is loaded into it. We do that for not showing a blank view when a banner failed to load. In the log, the first run of `ViewabilityTracker` returns some errors, and starting the second run it returns 0 (success).

Could that approach lead to lower eCPM?
